# Car Sales Analysis

Import the libraries we are going to use 

In [1]:
# importing the libraries we are going to use
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Step 1: Data Wrangling

Sourcing the data in the below cells. 

In [2]:
df = pd.read_csv("Car_sales.csv")

Viewing the First 10 and last 10 records of the data

In [3]:
df.head(10)


Manufacturer    Model  Sales_in_thousands  __year_resale_value Vehicle_type  \
0        Acura  Integra              16.919               16.360    Passenger   
1        Acura       TL              39.384               19.875    Passenger   
2        Acura       CL              14.114               18.225    Passenger   
3        Acura       RL               8.588               29.725    Passenger   
4         Audi       A4              20.397               22.255    Passenger   
5         Audi       A6              18.780               23.555    Passenger   
6         Audi       A8               1.380               39.000    Passenger   
7          BMW     323i              19.747                  NaN    Passenger   
8          BMW     328i               9.231               28.675    Passenger   
9          BMW     528i              17.527               36.125    Passenger   

   Price_in_thousands  Engine_size  Horsepower  Wheelbase  Width  Length  \
0               21.50          1.8       140.0      101.2   67.3   172.4   
1               28.40          3.2       225.0      108.1   70.3   192.9   
2                 NaN          3.2       225.0      106.9   70.6   192.0   
3               42.00          3.5       210.0      114.6   71.4   196.6   
4               23.99          1.8       150.0      102.6   68.2   178.0   
5               33.95          2.8       200.0      108.7   76.1   192.0   
6               62.00          4.2       310.0      113.0   74.0   198.2   
7               26.99          2.5       170.0      107.3   68.4   176.0   
8               33.40          2.8       193.0      107.3   68.5   176.0   
9               38.90          2.8       193.0      111.4   70.9   188.0   

   Curb_weight  Fuel_capacity  Fuel_efficiency Latest_Launch  \
0        2.639           13.2             28.0      2/2/2012   
1        3.517           17.2             25.0      6/3/2011   
2        3.470           17.2             26.0      1/4/2012   
3        3.850           18.0             22.0     3/10/2011   
4        2.998           16.4             27.0     10/8/2011   
5        3.561           18.5             22.0      8/9/2011   
6        3.902           23.7             21.0     2/27/2012   
7        3.179           16.6             26.0     6/28/2011   
8        3.197           16.6             24.0     1/29/2012   
9        3.472           18.5             25.0      4/4/2011   

   Power_perf_factor  
0          58.280150  
1          91.370778  
2                NaN  
3          91.389779  
4          62.777639  
5          84.565105  
6         134.656858  
7          71.191207  
8          81.877069  
9          83.998724

In [4]:
df.tail(10)

Manufacturer   Model  Sales_in_thousands  __year_resale_value  \
147   Volkswagen  Passat              51.102               16.725   
148   Volkswagen  Cabrio               9.569               16.575   
149   Volkswagen     GTI               5.596               13.760   
150   Volkswagen  Beetle              49.463                  NaN   
151        Volvo     S40              16.957                  NaN   
152        Volvo     V40               3.545                  NaN   
153        Volvo     S70              15.245                  NaN   
154        Volvo     V70              17.531                  NaN   
155        Volvo     C70               3.493                  NaN   
156        Volvo     S80              18.969                  NaN   

    Vehicle_type  Price_in_thousands  Engine_size  Horsepower  Wheelbase  \
147    Passenger               21.20          1.8       150.0      106.4   
148    Passenger               19.99          2.0       115.0       97.4   
149    Passenger               17.50          2.0       115.0       98.9   
150    Passenger               15.90          2.0       115.0       98.9   
151    Passenger               23.40          1.9       160.0      100.5   
152    Passenger               24.40          1.9       160.0      100.5   
153    Passenger               27.50          2.4       168.0      104.9   
154    Passenger               28.80          2.4       168.0      104.9   
155    Passenger               45.50          2.3       236.0      104.9   
156    Passenger               36.00          2.9       201.0      109.9   

     Width  Length  Curb_weight  Fuel_capacity  Fuel_efficiency Latest_Launch  \
147   68.5   184.1        3.043           16.4             27.0    10/30/2012   
148   66.7   160.4        3.079           13.7             26.0     5/31/2011   
149   68.3   163.3        2.762           14.6             26.0      4/1/2011   
150   67.9   161.1        2.769           14.5             26.0    10/20/2011   
151   67.6   176.6        2.998           15.8             25.0     2/18/2011   
152   67.6   176.6        3.042           15.8             25.0     9/21/2011   
153   69.3   185.9        3.208           17.9             25.0    11/24/2012   
154   69.3   186.2        3.259           17.9             25.0     6/25/2011   
155   71.5   185.7        3.601           18.5             23.0     4/26/2011   
156   72.1   189.8        3.600           21.1             24.0    11/14/2011   

     Power_perf_factor  
147          61.701381  
148          48.907372  
149          47.946841  
150          47.329632  
151          66.113057  
152          66.498812  
153          70.654495  
154          71.155978  
155         101.623357  
156          85.735655

Getting the general characteristics of the data at hand. 


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Manufacturer         157 non-null    object 
 1   Model                157 non-null    object 
 2   Sales_in_thousands   157 non-null    float64
 3   __year_resale_value  121 non-null    float64
 4   Vehicle_type         157 non-null    object 
 5   Price_in_thousands   155 non-null    float64
 6   Engine_size          156 non-null    float64
 7   Horsepower           156 non-null    float64
 8   Wheelbase            156 non-null    float64
 9   Width                156 non-null    float64
 10  Length               156 non-null    float64
 11  Curb_weight          155 non-null    float64
 12  Fuel_capacity        156 non-null    float64
 13  Fuel_efficiency      154 non-null    float64
 14  Latest_Launch        157 non-null    object 
 15  Power_perf_factor    155 non-null    flo

We can see that the data has got 157 entries with 16 columns having the data attributes.
We can also obtain knowledge of the missing value columns and the attribute data types. 



In [6]:
# the cartesian product of the dataset 
df.size

2512

In [7]:
df.columns

Index(['Manufacturer', 'Model', 'Sales_in_thousands', '__year_resale_value',
       'Vehicle_type', 'Price_in_thousands', 'Engine_size', 'Horsepower',
       'Wheelbase', 'Width', 'Length', 'Curb_weight', 'Fuel_capacity',
       'Fuel_efficiency', 'Latest_Launch', 'Power_perf_factor'],
      dtype='object')

In [8]:
# We can get a feel of the data statistics we are most likely to be dealing with
df.describe()

Sales_in_thousands  __year_resale_value  Price_in_thousands  \
count          157.000000           121.000000          155.000000   
mean            52.998076            18.072975           27.390755   
std             68.029422            11.453384           14.351653   
min              0.110000             5.160000            9.235000   
25%             14.114000            11.260000           18.017500   
50%             29.450000            14.180000           22.799000   
75%             67.956000            19.875000           31.947500   
max            540.561000            67.550000           85.500000   

       Engine_size  Horsepower   Wheelbase       Width      Length  \
count   156.000000  156.000000  156.000000  156.000000  156.000000   
mean      3.060897  185.948718  107.487179   71.150000  187.343590   
std       1.044653   56.700321    7.641303    3.451872   13.431754   
min       1.000000   55.000000   92.600000   62.600000  149.400000   
25%       2.300000  149.500000  103.000000   68.400000  177.575000   
50%       3.000000  177.500000  107.000000   70.550000  187.900000   
75%       3.575000  215.000000  112.200000   73.425000  196.125000   
max       8.000000  450.000000  138.700000   79.900000  224.500000   

       Curb_weight  Fuel_capacity  Fuel_efficiency  Power_perf_factor  
count   155.000000     156.000000       154.000000         155.000000  
mean      3.378026      17.951923        23.844156          77.043591  
std       0.630502       3.887921         4.282706          25.142664  
min       1.895000      10.300000        15.000000          23.276272  
25%       2.971000      15.800000        21.000000          60.407707  
50%       3.342000      17.200000        24.000000          72.030917  
75%       3.799500      19.575000        26.000000          89.414878  
max       5.572000      32.000000        45.000000         188.144323

In [9]:
for i,v in enumerate(df.columns):
    print(i,v)

0 Manufacturer
1 Model
2 Sales_in_thousands
3 __year_resale_value
4 Vehicle_type
5 Price_in_thousands
6 Engine_size
7 Horsepower
8 Wheelbase
9 Width
10 Length
11 Curb_weight
12 Fuel_capacity
13 Fuel_efficiency
14 Latest_Launch
15 Power_perf_factor


In [10]:
df.dtypes

Manufacturer            object
Model                   object
Sales_in_thousands     float64
__year_resale_value    float64
Vehicle_type            object
Price_in_thousands     float64
Engine_size            float64
Horsepower             float64
Wheelbase              float64
Width                  float64
Length                 float64
Curb_weight            float64
Fuel_capacity          float64
Fuel_efficiency        float64
Latest_Launch           object
Power_perf_factor      float64
dtype: object

From the avove we can see that the year_resale_value is of float but should be of datetime. We can correct that as in the code cell below.

In [11]:
df['__year_resale_value'] = pd.to_datetime(df['__year_resale_value'])

In [12]:
# now if we check the data types the datatype is correct
df.dtypes

Manufacturer                   object
Model                          object
Sales_in_thousands            float64
__year_resale_value    datetime64[ns]
Vehicle_type                   object
Price_in_thousands            float64
Engine_size                   float64
Horsepower                    float64
Wheelbase                     float64
Width                         float64
Length                        float64
Curb_weight                   float64
Fuel_capacity                 float64
Fuel_efficiency               float64
Latest_Launch                  object
Power_perf_factor             float64
dtype: object

In [13]:
# Let us check for thee duplicates in our data
df.duplicated().sum()

0

We have got no duplicates in our data

We can take care of the missing values in our dataset 

In [14]:
df.isna()

Manufacturer  Model  Sales_in_thousands  __year_resale_value  \
0           False  False               False                False   
1           False  False               False                False   
2           False  False               False                False   
3           False  False               False                False   
4           False  False               False                False   
..            ...    ...                 ...                  ...   
152         False  False               False                 True   
153         False  False               False                 True   
154         False  False               False                 True   
155         False  False               False                 True   
156         False  False               False                 True   

     Vehicle_type  Price_in_thousands  Engine_size  Horsepower  Wheelbase  \
0           False               False        False       False      False   
1           False               False        False       False      False   
2           False                True        False       False      False   
3           False               False        False       False      False   
4           False               False        False       False      False   
..            ...                 ...          ...         ...        ...   
152         False               False        False       False      False   
153         False               False        False       False      False   
154         False               False        False       False      False   
155         False               False        False       False      False   
156         False               False        False       False      False   

     Width  Length  Curb_weight  Fuel_capacity  Fuel_efficiency  \
0    False   False        False          False            False   
1    False   False        False          False            False   
2    False   False        False          False            False   
3    False   False        False          False            False   
4    False   False        False          False            False   
..     ...     ...          ...            ...              ...   
152  False   False        False          False            False   
153  False   False        False          False            False   
154  False   False        False          False            False   
155  False   False        False          False            False   
156  False   False        False          False            False   

     Latest_Launch  Power_perf_factor  
0            False              False  
1            False              False  
2            False               True  
3            False              False  
4            False              False  
..             ...                ...  
152          False              False  
153          False              False  
154          False              False  
155          False              False  
156          False              False  

[157 rows x 16 columns]

In [15]:
df.isna().sum()

Manufacturer            0
Model                   0
Sales_in_thousands      0
__year_resale_value    36
Vehicle_type            0
Price_in_thousands      2
Engine_size             1
Horsepower              1
Wheelbase               1
Width                   1
Length                  1
Curb_weight             2
Fuel_capacity           1
Fuel_efficiency         3
Latest_Launch           0
Power_perf_factor       2
dtype: int64

In [16]:
# to check the percentage of missing data
missing_per = df.isna().sum()*100/len(df)
print(missing_per)

Manufacturer            0.000000
Model                   0.000000
Sales_in_thousands      0.000000
__year_resale_value    22.929936
Vehicle_type            0.000000
Price_in_thousands      1.273885
Engine_size             0.636943
Horsepower              0.636943
Wheelbase               0.636943
Width                   0.636943
Length                  0.636943
Curb_weight             1.273885
Fuel_capacity           0.636943
Fuel_efficiency         1.910828
Latest_Launch           0.000000
Power_perf_factor       1.273885
dtype: float64


In [17]:
df['__year_resale_value'].fillna(method = 'ffill', inplace = True)
Price_mean=df['Price_in_thousands'].mean()
df['Price_in_thousands'].fillna(Price_mean, inplace = True)
Engine_mean = df['Engine_size'].mean()
df['Engine_size'].fillna(Engine_mean, inplace= True)
horse_mean=df['Horsepower'].mean()
df['Horsepower'].fillna(horse_mean, inplace = True)
df['Fuel_efficiency'].fillna(method = 'bfill', inplace= True)
df['Wheelbase'].fillna(method = 'ffill', inplace=True)
df['Width'].fillna(method='ffill', inplace=True)
df['Length'].fillna(method='bfill', inplace=True)
curb_mean = df['Curb_weight'].mean()
df['Curb_weight'].fillna(curb_mean, inplace=True)
fuel_mean = df['Fuel_capacity'].mean()
df['Fuel_capacity'].fillna(fuel_mean, inplace=True)
df['Fuel_efficiency'].fillna(method='bfill', inplace=True)
df['Power_perf_factor'].fillna(method='ffill', inplace=True)

I have used various ways to illustrate how one can eliminate missing values in the dataset 

In [20]:
# we confirm that all the missing values are dealt with
df.isna().sum()

Manufacturer           0
Model                  0
Sales_in_thousands     0
__year_resale_value    0
Vehicle_type           0
Price_in_thousands     0
Engine_size            0
Horsepower             0
Wheelbase              0
Width                  0
Length                 0
Curb_weight            0
Fuel_capacity          0
Fuel_efficiency        0
Latest_Launch          0
Power_perf_factor      0
dtype: int64